### PR TITLE
Adding webgl type and additional attributes to HTMLCanvasElement.getContext() fixes #5773

### PR DESCRIPTION
**Pull request by rohitverma007**
fixes microsoft/typescript#5773
This was long overdue, sorry about the delay :) 
Let me know if anything needs to be changed!

### CODE REVIEW

File: baselines/dom.generated.d.ts [hunk 1]
```diff
@@ -3761,9 +3762,9 @@ interface HTMLCanvasElement extends HTMLElement {
...
       * @param contextId The identifier (ID) of the type of canvas to create. Internet Explorer 9 and Internet Explorer 10 support only a 2-D context using canvas.getContext("2d"); IE11 Preview also supports 3-D or WebGL context using canvas.getContext("experimental-webgl");
       */
-    getContext(contextId: "2d"): CanvasRenderingContext2D;
-    getContext(contextId: "experimental-webgl"): WebGLRenderingContext;
-    getContext(contextId: string, ...args: any[]): CanvasRenderingContext2D | WebGLRenderingContext;
+    getContext(contextId: "2d", contextAttributes?: {alpha?: boolean; willReadFrequently?: boolean; storage?: boolean; [attrib: string]: boolean | string}): CanvasRenderingContext2D;
+    getContext(contextId: "webgl" | "experimental-webgl", contextAttributes: WebGLContextAttributes): WebGLRenderingContext;
+    getContext(contextId: string, contextAttributes?: {[prop: string]: boolean | string}): CanvasRenderingContext2D | WebGLRenderingContext;
```
Comment (mhegazy) [inline baselines/dom.generated.d.ts]: this seems a bit restrictive, and i could not find any documentation to support this. i would keep it to either `{}` or `any`

**Comment by mhegazy** [comment]
just one comment., other than that, looks good to me.

File: baselines/dom.generated.d.ts [hunk 2]
```diff
@@ -3761,9 +3762,9 @@ interface HTMLCanvasElement extends HTMLElement {
       * Returns an object that provides methods and properties for drawing and manipulating images and graphics on a canvas element in a document. A context object includes information about colors, line widths, fonts, and other graphic parameters that can be drawn on a canvas.
       * @param contextId The identifier (ID) of the type of canvas to create. Internet Explorer 9 and Internet Explorer 10 support only a 2-D context using canvas.getContext("2d"); IE11 Preview also supports 3-D or WebGL context using canvas.getContext("experimental-webgl");
       */
-    getContext(contextId: "2d"): CanvasRenderingContext2D;
-    getContext(contextId: "experimental-webgl"): WebGLRenderingContext;
-    getContext(contextId: string, ...args: any[]): CanvasRenderingContext2D | WebGLRenderingContext;
+    getContext(contextId: "2d", contextAttributes?: {alpha?: boolean; willReadFrequently?: boolean; storage?: boolean; [attrib: string]: boolean | string}): CanvasRenderingContext2D;
+    getContext(contextId: "webgl" | "experimental-webgl", contextAttributes: WebGLContextAttributes): WebGLRenderingContext;
```
Comment (zhengbli) [inline baselines/dom.generated.d.ts]: Why is the `contextAttributes` parameter not optional only in this overload?

File: baselines/dom.generated.d.ts (hunk 1 above)
Comment (rohitverma007) [inline baselines/dom.generated.d.ts]: @mhegazy are you referring to `{[prop: string]: boolean | string}` in line 3767 ?

File: baselines/dom.generated.d.ts (hunk 2 above)
Comment (rohitverma007) [inline baselines/dom.generated.d.ts]: ah no reason :) will fix it !

**Comment by rohitverma007** [comment]
updated

File: baselines/dom.generated.d.ts [hunk 3]
```diff
@@ -3763,9 +3764,9 @@ interface HTMLCanvasElement extends HTMLElement {
       * Returns an object that provides methods and properties for drawing and manipulating images and graphics on a canvas element in a document. A context object includes information about colors, line widths, fonts, and other graphic parameters that can be drawn on a canvas.
       * @param contextId The identifier (ID) of the type of canvas to create. Internet Explorer 9 and Internet Explorer 10 support only a 2-D context using canvas.getContext("2d"); IE11 Preview also supports 3-D or WebGL context using canvas.getContext("experimental-webgl");
       */
-    getContext(contextId: "2d"): CanvasRenderingContext2D;
-    getContext(contextId: "experimental-webgl"): WebGLRenderingContext;
-    getContext(contextId: string, ...args: any[]): CanvasRenderingContext2D | WebGLRenderingContext;
+    getContext(contextId: "2d", contextAttributes?: {alpha?: boolean; willReadFrequently?: boolean; storage?: boolean; [attrib: string]: boolean | string}): CanvasRenderingContext2D;
```
Comment (mhegazy) [inline baselines/dom.generated.d.ts]: do you mind creating a new interface WebGL2DContextAttributes  for these.

Comment (rohitverma007) [inline baselines/dom.generated.d.ts]: @mhegazy Sure, on it, although Canvas2DContextAttributes is probably a better name for it :)

**Comment by zhengbli** [comment]
ping @rohitverma007, is there an update on this one? Thanks!

**Comment by rohitverma007** [comment]
@zhengbli , my apologies! I have added the interface, please check the latest commit :)

File: baselines/webworker.generated.d.ts [hunk 4]
```diff
@@ -900,6 +900,13 @@ interface WorkerUtils extends Object, WindowBase64 {
     setTimeout(handler: any, timeout?: any, ...args: any[]): number;
 }
 
+interface Canvas2DContextAttributes {
```
Comment (mhegazy) [inline baselines/webworker.generated.d.ts]: hmm. @zhengbli this should not be here. is there a way to exclude it out from one output and not the other?

Comment (zhengbli) [inline baselines/webworker.generated.d.ts]: Yes, add a property `"flavor": "web"` should prevent it from being emitted to worker file

**Comment by rohitverma007** [comment]
@mhegazy @zhengbli fixed as requested

**Comment by mhegazy** [comment]
:+1:

File: inputfiles/addedTypes.json [hunk 5]
```diff
@@ -28,6 +28,29 @@
...
+            {
+                "name": "willReadFrequently?",
+                "type": "boolean"
+            },
+            {
+                "name": "storage?",
+                "type": "boolean"
+            },
+            {
+                "name": "[attrib: string]",
```
Comment (zhengbli) [inline inputfiles/addedTypes.json]: Better spell out the whole word, namely `attribute`.

**Comment by zhengbli** [comment]
@rohitverma007 Sorry for my delay, was busying with something else. Other than a minor comment, it looks fine to me. As soon as the conflict was resolved, I'll merge it in. Thank you!

**Comment by zhengbli** [comment]
👍 
Closing this and continue the work with resolved merge conflicts at #122.
Thanks!

**Pull request closed without merging**